ZeebeDbInconsistentException in column MESSAGE_PROCESSES_ACTIVE_BY_CORRELATION_KEY

## Description

I have a ZeebeProcessTest that fails and produces a `ZeebeDbInconsistentException` in the column `MESSAGE_PROCESSES_ACTIVE_BY_CORRELATION_KEY`. 

In the test, I deploy a process with a message start event and a message intermediate catch event. I publish two messages for the start event with the default message TTL and the same correlation key. As a result, it creates one new process instance. Then, I publish a message for the intermediate event. The command fails and the exception is written to the log.  

<details>
<summary>Full Exception</summary>

```
[zb-actors-5] ERROR io.camunda.zeebe.processor - Expected to process record 'LoggedEvent [position=31, key=-1, timestamp=1701768038850, sourceEventPosition=-1] RecordMetadata{recordType=COMMAND, valueType=MESSAGE, intent=PUBLISH, authorization=UNKNOWN}' successfully on stream processor, but caught unrecoverable exception.
[zb-actors-5] ERROR io.camunda.zeebe.logstreams - Actor StreamProcessor-1 failed in phase STARTED.
io.camunda.zeebe.db.ZeebeDbInconsistentException: Key DbCompositeKey{first=Start_Message_Flow, second=correlationKey1} in ColumnFamily MESSAGE_PROCESSES_ACTIVE_BY_CORRELATION_KEY already exists
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbColumnFamily.lambda$insert$1(InMemoryDbColumnFamily.java:66)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbColumnFamily.lambda$ensureInOpenTransaction$0(InMemoryDbColumnFamily.java:56)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbTransactionContext.runInTransaction(InMemoryDbTransactionContext.java:27)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbColumnFamily.ensureInOpenTransaction(InMemoryDbColumnFamily.java:55)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbColumnFamily.insert(InMemoryDbColumnFamily.java:61)
	at io.camunda.zeebe.engine.state.message.DbMessageState.putActiveProcessInstance(DbMessageState.java:279)
	at io.camunda.zeebe.engine.state.appliers.MessageStartEventSubscriptionCorrelatedApplier.applyState(MessageStartEventSubscriptionCorrelatedApplier.java:36)
	at io.camunda.zeebe.engine.state.appliers.MessageStartEventSubscriptionCorrelatedApplier.applyState(MessageStartEventSubscriptionCorrelatedApplier.java:16)
	at io.camunda.zeebe.engine.state.appliers.EventAppliers.applyState(EventAppliers.java:342)
	at io.camunda.zeebe.engine.processing.streamprocessor.writers.ResultBuilderBackedEventApplyingStateWriter.appendFollowUpEvent(ResultBuilderBackedEventApplyingStateWriter.java:56)
	at io.camunda.zeebe.engine.processing.streamprocessor.writers.ResultBuilderBackedEventApplyingStateWriter.appendFollowUpEvent(ResultBuilderBackedEventApplyingStateWriter.java:42)
	at io.camunda.zeebe.engine.processing.common.EventHandle.triggerMessageStartEvent(EventHandle.java:194)
	at io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnBufferedMessageStartEventBehavior.lambda$correlateNextBufferedMessage$0(BpmnBufferedMessageStartEventBehavior.java:82)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnBufferedMessageStartEventBehavior.correlateNextBufferedMessage(BpmnBufferedMessageStartEventBehavior.java:78)
	at io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnBufferedMessageStartEventBehavior.correlateMessage(BpmnBufferedMessageStartEventBehavior.java:66)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.lambda$getPostTransitionAction$10(ProcessProcessor.java:219)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.lambda$getPostTransitionAction$11(ProcessProcessor.java:217)
	at io.camunda.zeebe.util.Either$Right.ifRightOrLeft(Either.java:381)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.transitionTo(ProcessProcessor.java:189)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.onComplete(ProcessProcessor.java:73)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.onComplete(ProcessProcessor.java:27)
	at io.camunda.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processEvent(BpmnStreamProcessor.java:149)
	at io.camunda.zeebe.engine.processing.bpmn.BpmnStreamProcessor.lambda$processRecord$0(BpmnStreamProcessor.java:92)
	at io.camunda.zeebe.util.Either$Right.ifRightOrLeft(Either.java:381)
	at io.camunda.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processRecord(BpmnStreamProcessor.java:89)
	at io.camunda.zeebe.engine.Engine.process(Engine.java:127)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.batchProcessing(ProcessingStateMachine.java:360)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.lambda$processCommand$2(ProcessingStateMachine.java:275)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbTransaction.run(InMemoryDbTransaction.java:47)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.processCommand(ProcessingStateMachine.java:275)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.tryToReadNextRecord(ProcessingStateMachine.java:233)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.readNextRecord(ProcessingStateMachine.java:209)
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94)
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:47)
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:120)
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130)
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108)
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227)
```

</details>

<details>
<summary>Test output</summary>

```
[zb-actors-5] ERROR io.camunda.zeebe.processor - Expected to process record 'LoggedEvent [position=31, key=-1, timestamp=1701768038850, sourceEventPosition=-1] RecordMetadata{recordType=COMMAND, valueType=MESSAGE, intent=PUBLISH, authorization=UNKNOWN}' successfully on stream processor, but caught unrecoverable exception.
[zb-actors-5] ERROR io.camunda.zeebe.logstreams - Actor StreamProcessor-1 failed in phase STARTED.
io.camunda.zeebe.db.ZeebeDbInconsistentException: Key DbCompositeKey{first=Start_Message_Flow, second=correlationKey1} in ColumnFamily MESSAGE_PROCESSES_ACTIVE_BY_CORRELATION_KEY already exists
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbColumnFamily.lambda$insert$1(InMemoryDbColumnFamily.java:66)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbColumnFamily.lambda$ensureInOpenTransaction$0(InMemoryDbColumnFamily.java:56)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbTransactionContext.runInTransaction(InMemoryDbTransactionContext.java:27)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbColumnFamily.ensureInOpenTransaction(InMemoryDbColumnFamily.java:55)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbColumnFamily.insert(InMemoryDbColumnFamily.java:61)
	at io.camunda.zeebe.engine.state.message.DbMessageState.putActiveProcessInstance(DbMessageState.java:279)
	at io.camunda.zeebe.engine.state.appliers.MessageStartEventSubscriptionCorrelatedApplier.applyState(MessageStartEventSubscriptionCorrelatedApplier.java:36)
	at io.camunda.zeebe.engine.state.appliers.MessageStartEventSubscriptionCorrelatedApplier.applyState(MessageStartEventSubscriptionCorrelatedApplier.java:16)
	at io.camunda.zeebe.engine.state.appliers.EventAppliers.applyState(EventAppliers.java:342)
	at io.camunda.zeebe.engine.processing.streamprocessor.writers.ResultBuilderBackedEventApplyingStateWriter.appendFollowUpEvent(ResultBuilderBackedEventApplyingStateWriter.java:56)
	at io.camunda.zeebe.engine.processing.streamprocessor.writers.ResultBuilderBackedEventApplyingStateWriter.appendFollowUpEvent(ResultBuilderBackedEventApplyingStateWriter.java:42)
	at io.camunda.zeebe.engine.processing.common.EventHandle.triggerMessageStartEvent(EventHandle.java:194)
	at io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnBufferedMessageStartEventBehavior.lambda$correlateNextBufferedMessage$0(BpmnBufferedMessageStartEventBehavior.java:82)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnBufferedMessageStartEventBehavior.correlateNextBufferedMessage(BpmnBufferedMessageStartEventBehavior.java:78)
	at io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnBufferedMessageStartEventBehavior.correlateMessage(BpmnBufferedMessageStartEventBehavior.java:66)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.lambda$getPostTransitionAction$10(ProcessProcessor.java:219)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.lambda$getPostTransitionAction$11(ProcessProcessor.java:217)
	at io.camunda.zeebe.util.Either$Right.ifRightOrLeft(Either.java:381)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.transitionTo(ProcessProcessor.java:189)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.onComplete(ProcessProcessor.java:73)
	at io.camunda.zeebe.engine.processing.bpmn.container.ProcessProcessor.onComplete(ProcessProcessor.java:27)
	at io.camunda.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processEvent(BpmnStreamProcessor.java:149)
	at io.camunda.zeebe.engine.processing.bpmn.BpmnStreamProcessor.lambda$processRecord$0(BpmnStreamProcessor.java:92)
	at io.camunda.zeebe.util.Either$Right.ifRightOrLeft(Either.java:381)
	at io.camunda.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processRecord(BpmnStreamProcessor.java:89)
	at io.camunda.zeebe.engine.Engine.process(Engine.java:127)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.batchProcessing(ProcessingStateMachine.java:360)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.lambda$processCommand$2(ProcessingStateMachine.java:275)
	at io.camunda.zeebe.process.test.engine.db.InMemoryDbTransaction.run(InMemoryDbTransaction.java:47)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.processCommand(ProcessingStateMachine.java:275)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.tryToReadNextRecord(ProcessingStateMachine.java:233)
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.readNextRecord(ProcessingStateMachine.java:209)
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94)
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:47)
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:120)
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130)
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108)
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227)
[zb-actors-9] INFO io.camunda.zeebe.logstreams - Close appender for log stream null
[zb-actors-9] INFO io.camunda.zeebe.logstreams.impl.log.Sequencer - Closing sequencer for writing
[zb-actors-9] INFO io.camunda.zeebe.logstreams - On closing logstream null close 2 readers
[main] ERROR io.camunda.zeebe.process.test.extension.ZeebeProcessTestExtension - ===== Test failed!
[main] INFO io.camunda.zeebe.process.test.filters.logger.IncidentLogger - 
[main] INFO io.camunda.zeebe.process.test.filters.logger.RecordStreamLogger - 
The following records have been recorded during this test:
| COMMAND             DEPLOYMENT                         CREATE                        | (Processes: [MessageFlow.bpmn])
| EVENT               PROCESS                            CREATED                       | (Process: MessageFlow.bpmn)
| EVENT               DEPLOYMENT                         CREATED                       | 
| EVENT               MESSAGE_START_EVENT_SUBSCRIPTION   CREATED                       | (Process id: Start_Message_Flow), (Start event id: Start_Message), (Message name: Start_Message), (Correlation key: )
| COMMAND             MESSAGE                            PUBLISH                       | (Message name: Start_Message), (Correlation key: correlationKey1), (Variables: [Correlation_Key -> correlationKey1])
| EVENT               MESSAGE                            PUBLISHED                     | (Message name: Start_Message), (Correlation key: correlationKey1), (Variables: [Correlation_Key -> correlationKey1])
| EVENT               MESSAGE_START_EVENT_SUBSCRIPTION   CORRELATED                    | (Process id: Start_Message_Flow), (Start event id: Start_Message), (Message name: Start_Message), (Correlation key: correlationKey1)
| EVENT               PROCESS_EVENT                      TRIGGERING                    | (Target element id: Start_Message), (Variables: [Correlation_Key -> correlationKey1])
| COMMAND             PROCESS_INSTANCE                   ACTIVATE_ELEMENT              | (Element id: Start_Message_Flow), (Element type: PROCESS), (Event type: UNSPECIFIED), (Process id: Start_Message_Flow)
| EVENT               PROCESS_INSTANCE                   ELEMENT_ACTIVATING            | (Element id: Start_Message_Flow), (Element type: PROCESS), (Event type: UNSPECIFIED), (Process id: Start_Message_Flow)
| EVENT               PROCESS_INSTANCE                   ELEMENT_ACTIVATED             | (Element id: Start_Message_Flow), (Element type: PROCESS), (Event type: UNSPECIFIED), (Process id: Start_Message_Flow)
| EVENT               PROCESS_EVENT                      TRIGGERED                     | (Target element id: Start_Message), 
| EVENT               PROCESS_INSTANCE                   ELEMENT_ACTIVATING            | (Element id: Start_Message), (Element type: START_EVENT), (Event type: MESSAGE), (Process id: Start_Message_Flow)
| EVENT               PROCESS_INSTANCE                   ELEMENT_ACTIVATED             | (Element id: Start_Message), (Element type: START_EVENT), (Event type: MESSAGE), (Process id: Start_Message_Flow)
| EVENT               VARIABLE                           CREATED                       | (Name: Correlation_Key), (Value: "correlationKey1")
| COMMAND             PROCESS_INSTANCE                   COMPLETE_ELEMENT              | (Element id: Start_Message), (Element type: START_EVENT), (Event type: MESSAGE), (Process id: Start_Message_Flow)
| EVENT               PROCESS_INSTANCE                   ELEMENT_COMPLETING            | (Element id: Start_Message), (Element type: START_EVENT), (Event type: MESSAGE), (Process id: Start_Message_Flow)
| EVENT               VARIABLE                           CREATED                       | (Name: Correlation_Key), (Value: "correlationKey1")
| EVENT               PROCESS_INSTANCE                   ELEMENT_COMPLETED             | (Element id: Start_Message), (Element type: START_EVENT), (Event type: MESSAGE), (Process id: Start_Message_Flow)
| EVENT               PROCESS_INSTANCE                   SEQUENCE_FLOW_TAKEN           | (Element id: Flow_03wx9u1), (Element type: SEQUENCE_FLOW), (Event type: UNSPECIFIED), (Process id: Start_Message_Flow)
| COMMAND             PROCESS_INSTANCE                   ACTIVATE_ELEMENT              | (Element id: Intermediate_Message), (Element type: INTERMEDIATE_CATCH_EVENT), (Event type: MESSAGE), (Process id: Start_Message_Flow)
| EVENT               PROCESS_INSTANCE                   ELEMENT_ACTIVATING            | (Element id: Intermediate_Message), (Element type: INTERMEDIATE_CATCH_EVENT), (Event type: MESSAGE), (Process id: Start_Message_Flow)
| EVENT               PROCESS_MESSAGE_SUBSCRIPTION       CREATING                      | (Message name: Intermediate_Message), (Correlation key: correlationKey1), (Element id: Intermediate_Message), 
| COMMAND             MESSAGE_SUBSCRIPTION               CREATE                        | (Message name: Intermediate_Message), (Correlation key: correlationKey1), 
| EVENT               PROCESS_INSTANCE                   ELEMENT_ACTIVATED             | (Element id: Intermediate_Message), (Element type: INTERMEDIATE_CATCH_EVENT), (Event type: MESSAGE), (Process id: Start_Message_Flow)
| EVENT               MESSAGE_SUBSCRIPTION               CREATED                       | (Message name: Intermediate_Message), (Correlation key: correlationKey1), 
| COMMAND             PROCESS_MESSAGE_SUBSCRIPTION       CREATE                        | (Message name: Intermediate_Message), 
| EVENT               PROCESS_MESSAGE_SUBSCRIPTION       CREATED                       | (Message name: Intermediate_Message), (Correlation key: correlationKey1), (Element id: Intermediate_Message), 
| COMMAND             MESSAGE                            PUBLISH                       | (Message name: Start_Message), (Correlation key: correlationKey1), (Variables: [Correlation_Key -> correlationKey1])
| EVENT               MESSAGE                            PUBLISHED                     | (Message name: Start_Message), (Correlation key: correlationKey1), (Variables: [Correlation_Key -> correlationKey1])
| COMMAND             MESSAGE                            PUBLISH                       | (Message name: Intermediate_Message), (Correlation key: correlationKey1), 

io.camunda.zeebe.client.api.command.ClientStatusException: deadline exceeded after 9.999812546s. Name resolution delay 0.000000000 seconds. [closed=[], open=[[remote_addr=/127.0.0.1:45075]]]

	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.transformExecutionException(ZeebeClientFutureImpl.java:116)
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:54)
	at io.camunda.zeebe.process.test.examples.MessageTestsSmall.uniqueCorrelationKeyTest(MessageTestsSmall.java:69)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 9.999812546s. Name resolution delay 0.000000000 seconds. [closed=[], open=[[remote_addr=/127.0.0.1:45075]]]
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:52)
	... 4 more
Caused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 9.999812546s. Name resolution delay 0.000000000 seconds. [closed=[], open=[[remote_addr=/127.0.0.1:45075]]]
	at io.grpc.Status.asRuntimeException(Status.java:537)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:481)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:574)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:72)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:742)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

</details>

### Expected behaviour

The message for the intermediate event can be published successfully. After the first process instance is completed, a second process instance is created for the second buffered message. 

## Reproduction steps

<!-- How can we reproduce your issue? -->
<details>
<summary>Test code</summary>
<!-- If possible, please add your test code in this code block. -->

```
package org.camunda.support;

import java.time.Duration;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import io.camunda.zeebe.client.ZeebeClient;
import io.camunda.zeebe.process.test.api.ZeebeTestEngine;
import io.camunda.zeebe.process.test.extension.ZeebeProcessTest;
import io.camunda.zeebe.process.test.filters.RecordStream;


@ZeebeProcessTest
class MessageTestsSmall {
    private ZeebeTestEngine engine;
    private ZeebeClient client;
    private RecordStream recordStream;

    @BeforeEach
    void deployProcesses() {
        client.newDeployResourceCommand()
                .addResourceFromClasspath("MessageFlow.bpmn")
                .send()
                .join();
    }

    @Test
    void uniqueCorrelationKeyTest() throws InterruptedException {
      
    	String correlationKey1 = "correlationKey1";
        String correlationKey = "Correlation_Key";

        client.newPublishMessageCommand()
                .messageName("Start_Message")
                .correlationKey(correlationKey1)
               // .timeToLive(Duration.ofSeconds(1))
                .variable(correlationKey, correlationKey1)
                .send().join();

        client.newPublishMessageCommand()
                .messageName("Start_Message")
                .correlationKey(correlationKey1)
             //   .timeToLive(Duration.ofSeconds(1))
                .variable(correlationKey, correlationKey1)
                .send().join();


        client.newPublishMessageCommand()
                .messageName("Intermediate_Message")
                .correlationKey(correlationKey1)
                .timeToLive(Duration.ofMinutes(1))
                .send().join();
    }

}
```

</details>

<details>
<summary>Process</summary>

<!-- If possible, please add your process in this code block. -->

```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" xmlns:modeler="http://camunda.org/schema/modeler/1.0" id="Definitions_1e66ubr" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.15.0" modeler:executionPlatform="Camunda Cloud" modeler:executionPlatformVersion="8.1.0">
  <bpmn:process id="Start_Message_Flow" name="StartMessageFlow" isExecutable="true">
    <bpmn:endEvent id="End_Event" name="End">
      <bpmn:incoming>Flow_1oo7035</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:startEvent id="Start_Message" name="StartMessage">
      <bpmn:outgoing>Flow_03wx9u1</bpmn:outgoing>
      <bpmn:messageEventDefinition id="MessageEventDefinition_0vf3f3f" messageRef="Message_0klcjh4" />
    </bpmn:startEvent>
    <bpmn:intermediateCatchEvent id="Intermediate_Message" name="IntermediateMessage">
      <bpmn:incoming>Flow_03wx9u1</bpmn:incoming>
      <bpmn:outgoing>Flow_1oo7035</bpmn:outgoing>
      <bpmn:messageEventDefinition id="MessageEventDefinition_1sfcp0s" messageRef="Message_0u4d2gn" />
    </bpmn:intermediateCatchEvent>
    <bpmn:sequenceFlow id="Flow_03wx9u1" sourceRef="Start_Message" targetRef="Intermediate_Message" />
    <bpmn:sequenceFlow id="Flow_1oo7035" sourceRef="Intermediate_Message" targetRef="End_Event" />
  </bpmn:process>
  <bpmn:message id="Message_3ob5mlf" name="Message_Reference">
    <bpmn:extensionElements>
      <zeebe:subscription correlationKey="=Correlation_Key" />
    </bpmn:extensionElements>
  </bpmn:message>
  <bpmn:message id="Message_0klcjh4" name="Start_Message" />
  <bpmn:message id="Message_0u4d2gn" name="Intermediate_Message">
    <bpmn:extensionElements>
      <zeebe:subscription correlationKey="=Correlation_Key" />
    </bpmn:extensionElements>
  </bpmn:message>
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Start_Message_Flow">
      <bpmndi:BPMNShape id="Event_0jkmmby_di" bpmnElement="End_Event">
        <dc:Bounds x="522" y="99" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="530" y="142" width="20" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_09qd40m_di" bpmnElement="Start_Message">
        <dc:Bounds x="152" y="99" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="136" y="142" width="69" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_03kffle_di" bpmnElement="Intermediate_Message">
        <dc:Bounds x="322" y="99" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="299" y="142" width="83" height="27" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="Flow_03wx9u1_di" bpmnElement="Flow_03wx9u1">
        <di:waypoint x="188" y="117" />
        <di:waypoint x="322" y="117" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1oo7035_di" bpmnElement="Flow_1oo7035">
        <di:waypoint x="358" y="117" />
        <di:waypoint x="522" y="117" />
      </bpmndi:BPMNEdge>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>

```

</details>

![Screenshot from 2023-12-05 10-35-27](https://github.com/camunda/zeebe-process-test/assets/4305769/fd46fb6e-c940-4729-9072-919e1bcd4587)

## Environment

- OS: <!-- e.g. Linux -->
- Version: `8.4.0-SNAPSHOT`

The issue was reported via [Support](https://jira.camunda.com/browse/SUPPORT-19378).

--- a/engine/src/main/java/io/camunda/zeebe/process/test/engine/db/InMemoryDbTransaction.java
+++ b/engine/src/main/java/io/camunda/zeebe/process/test/engine/db/InMemoryDbTransaction.java
@@ -73,6 +73,10 @@ final class InMemoryDbTransaction implements ZeebeDbTransaction, InMemoryDbState
   public byte[] get(final FullyQualifiedKey fullyQualifiedKey) {
     final Bytes key = fullyQualifiedKey.getKeyBytes();
 
+    if (deletedKeys.contains(key)) {
+      return null;
+    }
+
     final Bytes valueInCache = transactionCache.get(key);
 
     if (valueInCache != null) {

--- a/engine/src/main/java/io/camunda/zeebe/process/test/engine/db/InMemoryDbTransaction.java
+++ b/engine/src/main/java/io/camunda/zeebe/process/test/engine/db/InMemoryDbTransaction.java
@@ -108,6 +108,7 @@ final class InMemoryDbTransaction implements ZeebeDbTransaction, InMemoryDbState
   @Override
   public boolean contains(final FullyQualifiedKey fullyQualifiedKey) {
     final Bytes keyBytes = fullyQualifiedKey.getKeyBytes();
-    return transactionCache.containsKey(keyBytes) || database.containsKey(keyBytes);
+    return !deletedKeys.contains(keyBytes)
+        && (transactionCache.containsKey(keyBytes) || database.containsKey(keyBytes));
   }
 }

--- a/engine/src/test/java/io/camunda/zeebe/process/test/engine/db/InMemoryZeebeDbTransactionTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/process/test/engine/db/InMemoryZeebeDbTransactionTest.java
@@ -441,6 +441,26 @@ public final class InMemoryZeebeDbTransactionTest {
   }
 
   @Test
+  void shouldAllowDeleteAndInsertInTransaction() throws Exception {
+    // given
+    oneKey.wrapLong(1);
+    oneValue.wrapLong(-1L);
+    twoValue.wrapLong(-2L);
+    oneColumnFamily.insert(oneKey, oneValue);
+    transactionContext.getCurrentTransaction().commit();
+
+    // when
+    transactionContext.runInTransaction(
+        () -> {
+          oneColumnFamily.deleteExisting(oneKey);
+          oneColumnFamily.insert(oneKey, twoValue);
+        });
+
+    // then
+    assertThat(oneColumnFamily.get(oneKey).getValue()).isEqualTo(twoValue.getValue());
+  }
+
+  @Test
   void shouldNotCommitOnError() {
     // given
     oneKey.wrapLong(1);

--- a/engine/src/test/java/io/camunda/zeebe/process/test/engine/db/InMemoryZeebeDbTransactionTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/process/test/engine/db/InMemoryZeebeDbTransactionTest.java
@@ -461,6 +461,29 @@ public final class InMemoryZeebeDbTransactionTest {
   }
 
   @Test
+  void shouldNotGetByKeyIfDeletedInTransaction() throws Exception {
+    // given
+    oneKey.wrapLong(1);
+    oneValue.wrapLong(-1L);
+    twoValue.wrapLong(-2L);
+    oneColumnFamily.insert(oneKey, oneValue);
+    transactionContext.getCurrentTransaction().commit();
+
+    // when
+    transactionContext.runInTransaction(
+        () -> {
+          oneColumnFamily.deleteExisting(oneKey);
+
+          if (oneColumnFamily.get(oneKey) != null) {
+            fail("Should not be able to get deleted key.");
+          }
+        });
+
+    // then
+    assertThat(oneColumnFamily.get(oneKey)).isNull();
+  }
+
+  @Test
   void shouldNotCommitOnError() {
     // given
     oneKey.wrapLong(1);

